fax/implicit/twophase.py

```diff
@@ -32,98 +29,44 @@ def convergence_test(x_new, x_old):
-            init_x=init_dxs,
-            func=dfp_fn,
-            convergence_test=convergence_test,
-            max_iter=default_max_iter,
-            batched_iter_size=default_batched_iter_size,
-        )
 
-        return fp_vjp_fn(dsol.value)[1], dsol
-
-    return _adjoint_iteration_vjp
-
-
-def two_phase_solver(param_func, forward_solver=None, default_rtol=1e-4,
-                     default_atol=1e-4, default_max_iter=5000,
-                     default_batched_iter_size=1):
-    """ Create an implicit function of the parameters and define its VJP rule.
-
-    Args:
-        param_func: A "parametric" operator (i.e., callable) taking in some
-            parameters and returning a function for which we seek a fixed point.
-        forward_solver:
-        default_rtol (float, optional): The relative tolerance (as used by `np.isclose`).
-            Defaults to 1e-4.
-        default_atol (float, optional): The absolute tolerance (as used by `np.isclose`).
-            Defaults to 1e-4.
-        default_max_iter (int, optional): The maximum number of iterations. Defaults to 5000.
-        default_batched_iter_size (int, optional):  The number of iterations to be
-            unrolled and executed per iterations of `while_loop` op. Defaults to 1.
-
-    Returns:
-        callable: Binary callable with signature ``f(x_0, params)`` returning a solution ``x``
-            such that ``param_func(params)(x) = x``. The returned callable has a registered
-            VJP rule so that it can be composed with other functions in and end-to-end fashion.
-    """
-    # if no solver is specified, create a default solver
-    if forward_solver is None:
-        forward_solver = make_forward_fixed_point_iteration(
-            param_func, default_rtol=default_rtol, default_atol=default_atol,
-            default_max_iter=default_max_iter, default_batched_iter_size=default_batched_iter_size)
-
-    adjoint_iteration_vjp = make_adjoint_fixed_point_iteration(
-        param_func, default_rtol=default_rtol, default_atol=default_atol,
-        default_max_iter=default_max_iter, default_batched_iter_size=default_batched_iter_size)
-
-    @jax.custom_vjp
-    def two_phase_op(init_xs, params):
-        return forward_solver(init_xs, params)
-
-    def two_phase_fwd(init_xs, params):
-        ans = two_phase_op(init_xs, params)
-        return ans, (ans, init_xs, params)
-
-    def two_phase_bwd(res, g):
-        ans, init_xs, params = res
-        dvalue, dconverged, diter, dprev_value = g
-        # these tensors are returned only for monitoring and have no defined gradient
-        del dconverged, diter, dprev_value
-        zeroed_xs = tree_util.tree_map(np.zeros_like, init_xs)
-        g_times_d_params = adjoint_iteration_vjp(dvalue, ans, init_xs, params)[0]
-        return zeroed_xs, g_times_d_params
-
-    two_phase_op.defvjp(two_phase_fwd, two_phase_bwd)
-
-    return two_phase_op
+two_phase_solve.defvjp(two_phase_fwd, two_phase_rev)
```

Comment:
Awesome how much simpler this got!

Reply:
Thanks! I got jealous of the simplicity of the fixed point example from the `jax` docs. The new `custom_vjp` api makes a simple implementation much easier to pull off.